clippy_lints/src/default_box_assignments.rs

```diff
@@ -0,0 +1,87 @@
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::is_default_equivalent_call;
+use clippy_utils::source::snippet;
+use clippy_utils::ty::implements_trait;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind, LangItem};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::{self, GenericArgKind, Ty};
+use rustc_session::declare_lint_pass;
+use rustc_span::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Detects assignments of `Default::default()` to a place of type `Box<T>`.
+    ///
+    /// ### Why is this bad?
+    /// This incurs an extra heap allocation compared to assigning the boxed
+    /// storage.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// let mut b = Box::new(1u32);
+    /// b = Default::default();
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let mut b = Box::new(1u32);
+    /// *b = Default::default();
+    /// ```
+    #[clippy::version = "1.89.0"]
+    pub DEFAULT_BOX_ASSIGNMENTS,
+    perf,
+    "assigning `Default::default()` to `Box<T>` is inefficient"
+}
+declare_lint_pass!(DefaultBoxAssignments => [DEFAULT_BOX_ASSIGNMENTS]);
+
+impl LateLintPass<'_> for DefaultBoxAssignments {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &'_ Expr<'_>) {
+        if let ExprKind::Assign(lhs, rhs, _) = &expr.kind {
+            let lhs_ty = cx.typeck_results().expr_ty(lhs);
+            if is_box_of_default(lhs_ty, cx) && is_default_call(rhs, cx) {
+                span_lint_and_then(
+                    cx,
+                    DEFAULT_BOX_ASSIGNMENTS,
+                    expr.span,
+                    "assigning `Default::default()` to `Box<T>`",
```

Comment:
Here, `<T>` doesn't correspond to anything concrete for the user. Maybe something like "creating a new `Box` with default content" would be clearer, with a "replace existing content with default instead" suggestion message.